File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        ListNode temp = head;
        int n = 0;

        while (temp != null) {
            temp = temp.next;
            n++;
        }

        if (n == 1) return null;
        ListNode middleNode = head;

        for (int x = 0; x < n/ 2 - 1; x++) middleNode = middleNode.next;

        ListNode prev = middleNode;
        middleNode = middleNode.next;
        if (middleNode != null) {
            prev.next = middleNode.next;
            middleNode.next = null;
        }
        else prev = middleNode;

        return head;
    }
}

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        temp, n = head, 0
        while temp is not None :
            temp = temp.next
            n += 1

        if n == 1 : return None
        # print(n // 2)
        # This is same as deleting a node at an index in a list.
        middle_node = head
        for x in range(n // 2 - 1) :
            middle_node = middle_node.next

        prev = middle_node
        middle_node = middle_node.next
        if middle_node is not None : 
            prev.next = middle_node.next
            middle_node.next = None
        
        else : prev.next = middle_node
        return head